pandas/core/reshape/merge.py

```diff
@@ -1583,6 +1608,25 @@ def _validate(self, validate: str) -> None:
             )
 
 
+def _semi_helper(
+    leftdf: DataFrame,
+    left: Index | DataFrame,
+    right: Index | DataFrame,
+) -> DataFrame:
+    if not isinstance(left, Index):
+        if len(left.columns) == 1:
+            left = Index(left.values.flatten())
+        else:
+            left = MultiIndex.from_frame(left)
+    if not isinstance(right, Index):
+        if len(right.columns) == 1:
+            right = Index(right.values.flatten())
+        else:
+            right = MultiIndex.from_frame(right)
+    subset = left.isin(right)
```

Comment:
What's the performance of this? Say you have two large frames (maybe check 10k and 100k rows) that have the same result whether it be a left or leftsemi (I think e.g. the keys being unique guarantees this), how does the performance compare?

Reply:
> that have the same result whether it be a left or leftsemi

I will have to check that, but doesn't `leftsemi` just return the left df, nothing from the right df?

Reply:
Ah, when I said "same result" above, I was merely thinking that the left join won't produce duplicate rows (so the results are the same length). You're correct the results won't be the same.

Reply:
```python
left = pd.DataFrame({"a":range(100000), "b":["a","b","c","d"]*25000})
right = pd.DataFrame({"a":range(100000), "b":["d","e","f", "g"]*25000})
%timeit left.merge(right)
41 ms ± 321 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit left.merge(right, how="leftsemi")
63.9 ms ± 2.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Reply:
Thanks! `merge` will use all columns that are in common by default, your example is an empty result. While informative, I think a non-empty result is probably more common for testing.

I'm seeing this is faster (for this particular example which _may_ be too simple) on one column, and ~5x slower once you get to three or more. While there may be opportunities for better performance, that seems reasonable for a new feature.

<details>
<summary>Code</summary>

```
for n_columns in (1, 2, 4, 8, 16, 32, 64):
    on = list(range(n_columns))
    left = pd.DataFrame(
        {
            **{i: range(100000) for i in on},
            **{"values":["a","b","c","d"]*25000},
        }
    )
    right = pd.DataFrame(
        {
            **{i: range(100000) for i in on},
            **{"values":["d","e","f", "g"]*25000},
        }
    )

    print(f'{n_columns=}')
    %timeit left.merge(right, on=on)
    %timeit left.merge(right, on=on, how="leftsemi")
    print('---')
```

</details>

```
n_columns=1
6.69 ms ± 22.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.26 ms ± 2.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=2
8.5 ms ± 31 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
13.8 ms ± 229 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=4
16.3 ms ± 228 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
54.1 ms ± 127 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
---
n_columns=8
29.9 ms ± 165 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
108 ms ± 841 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
---
n_columns=16
58.3 ms ± 330 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
223 ms ± 1.74 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
---
n_columns=32
118 ms ± 237 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
478 ms ± 3.51 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
---
n_columns=64
240 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.02 s ± 5.97 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Any suggestions how performance can be improved?

Reply:
No, not offhand. Just to show there is room for performance improvement, here is a hacky solution (won't work in all cases):

```
def my_leftsemi(left, right, on):
    right = right.drop(columns='values')
    left['left_row'] = range(len(left))
    left['index'] = left.index
    result = (
        left.merge(right, on=on)
        .drop_duplicates(subset='left_row')
        .drop(columns='left_row')
    )
    result = result.set_index('index').sort_index()
    result.index.name = None
    return result
```

This produces the same result, and has runtime similar to the current left join on many columns:

```
n_columns=1
7.04 ms ± 59.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.12 ms ± 3.31 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10 ms ± 123 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=2
9.47 ms ± 81.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
13.9 ms ± 110 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.1 ms ± 74.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=4
17.4 ms ± 95.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
56 ms ± 514 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
19.8 ms ± 188 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=8
30.3 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
111 ms ± 1.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
36.8 ms ± 386 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
---
n_columns=16
57.7 ms ± 94.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
229 ms ± 2.97 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
72.3 ms ± 285 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
---
n_columns=32
116 ms ± 222 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
466 ms ± 2.94 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
133 ms ± 272 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
---
n_columns=64
235 ms ± 275 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
982 ms ± 3.22 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
260 ms ± 694 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Thanks @rhshadrach for the review.

I took a look into `leftsemi` and it is taking long because of the use of `isin` (I think that was suggested in a comment in the bug report).
We can implement `leftsemi` without using `isin` too (something similar to your `my_leftsemi` func). But I think that would get more complicated if we try to cover all the use cases. Let me know if you think I should take this PR towards that direction.

Digressing a little from above, I also tried another use case perf. I changed it a little (output lens are still same), and the perf diff has changed significantly in the opposite direction.

Let me know your thoughts.

<details>
<summary>Code</summary>

```python
import pandas as pd
for n_columns in (1, 2, 4, 8, 16, 32, 64):
    on = list(range(n_columns))
    left = pd.DataFrame(
        {
            **{"l"+str(i): range(100000) for i in on},
            **{"values_l":["a","b","c","d"]*25000},
            **{"key": range(2, 100002)}
        }
    )
    right = pd.DataFrame(
        {
            **{"r"+str(i): range(100000) for i in on},
            **{"values_r":["d","e","f", "g"]*25000},
            **{"key": range(100000)}
        }
    )
    print(f'{n_columns=}')
    %timeit left.merge(right)
    %timeit left.merge(right, how="leftsemi")
    print('---')
```
</details>



```python
n_columns=1
45.3 ms ± 6.44 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
7.82 ms ± 507 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=2
29.7 ms ± 2.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
7.75 ms ± 301 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=4
33.9 ms ± 2.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
8.48 ms ± 434 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=8
37.4 ms ± 2.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
9.41 ms ± 404 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=16
46.6 ms ± 2.22 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
11.8 ms ± 353 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=32
68 ms ± 3.17 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
15.8 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
---
n_columns=64
115 ms ± 19.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
26.1 ms ± 2.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Reply:
> We can implement `leftsemi` without using `isin` too (something similar to your `my_leftsemi` func). But I think that would get more complicated if we try to cover all the use cases. Let me know if you think I should take this PR towards that direction.

I don't think this should go in that direction unless there already is a viable alternative; to my knowledge there isn't. As I remarked above, while this performance is less than left-join, I feel it's sufficient for the feature.

> Digressing a little from above, I also tried another use case perf. I changed it a little (output lens are still same), and the perf diff has changed significantly in the opposite direction.

Left join has to take all the columns from right where as leftsemi does not. In other words, these operations in this case are quite different, so I don't think it's meaningful to compare. Also, perhaps I didn't highlight this well enough, but joining on 1-3 columns is either significantly faster or comparable to left joins. In my experience this is the common case; I can't remember the last time I joined on 10 columns :laughing:

Reply:
> Left join has to take all the columns from right where as leftsemi does not

Rational behind this test was, I think that's the use case where someone would use `leftsemi` join.
But yes, I do agree that since the output is widely different, this is not fair to `leftjoin`.